diff --git a/lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp b/lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp
index c3fe48f..e4bbd3d 100644
--- a/lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp
+++ b/lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp
@@ -706,7 +706,15 @@ void WhenOpVisitor::visitStmt(WhenOp whenOp) {
 
 // NOLINTNEXTLINE(misc-no-recursion)
 void WhenOpVisitor::visitStmt(LayerBlockOp layerBlockOp) {
+  auto savedAnds = createdLTLAndOps;
+  auto savedImpls = createdLTLImplicationOps;
+  auto savedClocks = createdLTLClockOps;
+
   process(*layerBlockOp.getBody());
+
+  createdLTLAndOps = std::move(savedAnds);
+  createdLTLImplicationOps = std::move(savedImpls);
+  createdLTLClockOps = std::move(savedClocks);
 }
 
 void WhenOpVisitor::visitStmt(RefForceOp op) {
diff --git a/test/Dialect/FIRRTL/expand-whens.mlir b/test/Dialect/FIRRTL/expand-whens.mlir
index 03c4148..8897e27 100644
--- a/test/Dialect/FIRRTL/expand-whens.mlir
+++ b/test/Dialect/FIRRTL/expand-whens.mlir
@@ -776,4 +776,31 @@ firrtl.module @instance_choice_test(in %a : !firrtl.uint<1>,
   }
 }
 
+
+firrtl.layer @Bar bind attributes {sym_visibility = "private"} {}
+
+// CHECK-LABEL: firrtl.module @Issue10104
+firrtl.module @Issue10104() {
+  %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  firrtl.when %c0_ui1 : !firrtl.uint<1> {
+    // CHECK: firrtl.layerblock @Bar {
+    firrtl.layerblock @Bar {
+      // CHECK-NEXT: [[COND_0:%.*]] = firrtl.int.ltl.implication %c0_ui1, %c0_ui1
+      %0 = firrtl.int.ltl.implication %c0_ui1, %c0_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
+      // CHECK-NEXT: [[AND_0:%.*]] = firrtl.int.ltl.and %c0_ui1, %c0_ui1
+      // CHECK-NEXT: [[IMPL_0:%.*]] = firrtl.int.ltl.implication [[AND_0]], %c0_ui1
+      // CHECK-NEXT: firrtl.int.verif.assert [[IMPL_0]], %c0_ui1
+      firrtl.int.verif.assert %0, %c0_ui1 : !firrtl.uint<1>, !firrtl.uint<1>
+    }
+    // CHECK: firrtl.layerblock @Bar {
+    firrtl.layerblock @Bar {
+      // CHECK-NEXT: [[COND_1:%.*]] = firrtl.int.ltl.implication %c0_ui1, %c0_ui1
+      %1 = firrtl.int.ltl.implication %c0_ui1, %c0_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
+      // CHECK-NEXT: [[AND_1:%.*]] = firrtl.int.ltl.and %c0_ui1, %c0_ui1
+      // CHECK-NEXT: [[IMPL_1:%.*]] = firrtl.int.ltl.implication [[AND_1]], %c0_ui1
+      // CHECK-NEXT: firrtl.int.verif.assert [[IMPL_1]], %c0_ui1
+      firrtl.int.verif.assert %1, %c0_ui1 : !firrtl.uint<1>, !firrtl.uint<1>
+    }
+  }
+}
 }
diff --git a/test/lit.cfg.py b/test/lit.cfg.py
index 727c07d..adc6ea3 100644
--- a/test/lit.cfg.py
+++ b/test/lit.cfg.py
@@ -19,7 +19,7 @@ from lit.llvm.subst import FindTool
 # name: The name of this test suite.
 config.name = 'CIRCT'
 
-config.test_format = lit.formats.ShTest(not llvm_config.use_lit_shell)
+config.test_format = lit.formats.ShTest()
 
 # suffixes: A list of file extensions to treat as test files.
 config.suffixes = [
